app/src/main/java/org/thoughtcrime/securesms/util/MemoryFileDescriptorProxy.java

```diff
@@ -0,0 +1,89 @@
+package org.thoughtcrime.securesms.util;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.MemoryFile;
+import android.os.ParcelFileDescriptor;
+import android.os.ProxyFileDescriptorCallback;
+import android.os.storage.StorageManager;
+import android.system.ErrnoException;
+import android.system.OsConstants;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import org.signal.core.util.logging.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+@RequiresApi(api = 26)
+final class MemoryFileDescriptorProxy {
+
+  private static final String TAG = Log.tag(MemoryFileDescriptorProxy.class);
+
+  public static ParcelFileDescriptor create(@NonNull Context context,
+                                            @NonNull MemoryFile file)
+      throws IOException
+  {
+    StorageManager storageManager = Objects.requireNonNull(context.getSystemService(StorageManager.class));
+    HandlerThread  thread         = new HandlerThread("MemoryFile");
+
+    thread.start();
+    Log.i(TAG, "Thread started");
+
+    Handler       handler       = new Handler(thread.getLooper());
+    ProxyCallback proxyCallback = new ProxyCallback(file, () -> {
+      if (thread.quitSafely()) {
+        Log.i(TAG, "Thread quitSafely true");
+      } else {
+        Log.w(TAG, "Thread quitSafely false");
+      }
+    });
+
+    ParcelFileDescriptor parcelFileDescriptor = storageManager.openProxyFileDescriptor(ParcelFileDescriptor.MODE_READ_ONLY,
+                                                                                       proxyCallback,
+                                                                                       handler);
+
+    Log.i(TAG, "Created");
+    return parcelFileDescriptor;
+  }
+
+  private static final class ProxyCallback extends ProxyFileDescriptorCallback {
+
+    private final MemoryFile memoryFile;
+    private final Runnable   onClose;
+
+    ProxyCallback(@NonNull MemoryFile memoryFile, Runnable onClose) {
+      this.memoryFile = memoryFile;
+      this.onClose    = onClose;
+    }
+
+    @Override
+    public long onGetSize() {
+      return memoryFile.length();
+    }
+
+    @Override
+    public int onRead(long offset, int size, byte[] data) throws ErrnoException {
+      try {
+        InputStream inputStream = memoryFile.getInputStream();
+        if(inputStream.skip(offset) != offset){
+          throw new AssertionError();
+        }
+        return inputStream.read(data, 0, size);
+      } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)